src/ImageSharp/Processing/Processors/Quantization/EuclideanPixelMap{TPixel,TCache}.cs

```diff
@@ -0,0 +1,219 @@
+    {
+        float deltaR = a.R - b.R;
+        float deltaG = a.G - b.G;
+        float deltaB = a.B - b.B;
+        float deltaA = a.A - b.A;
+        return (deltaR * deltaR) + (deltaG * deltaG) + (deltaB * deltaB) + (deltaA * deltaA);
+    }
+
+    /// <inheritdoc/>
+    public override void Dispose() => this.cache.Dispose();
+}
+
+/// <summary>
+/// Represents a map of colors to indices.
+/// </summary>
+/// <typeparam name="TPixel">The pixel format.</typeparam>
+internal abstract class PixelMap<TPixel> : IDisposable
+    where TPixel : unmanaged, IPixel<TPixel>
+{
+    /// <summary>
+    /// Gets the color palette of this <see cref="PixelMap{TPixel}"/>.
+    /// </summary>
+    public ReadOnlyMemory<TPixel> Palette { get; private protected set; }
+
+    /// <summary>
+    /// Returns the closest color in the palette and the index of that pixel.
+    /// </summary>
+    /// <param name="color">The color to match.</param>
+    /// <param name="match">The matched color.</param>
+    /// <returns>
+    /// The <see cref="int"/> index.
+    /// </returns>
+    public abstract int GetClosestColor(TPixel color, out TPixel match);
+
+    /// <summary>
+    /// Clears the map, resetting it to use the given palette.
+    /// </summary>
+    /// <param name="palette">The color palette to map from.</param>
+    public abstract void Clear(ReadOnlyMemory<TPixel> palette);
+
+    /// <summary>
+    /// Allows setting the transparent index after construction.
+    /// </summary>
+    /// <param name="index">An explicit index at which to match transparent pixels.</param>
+    public abstract void SetTransparentIndex(int index);
+
+    /// <inheritdoc/>
+    public abstract void Dispose();
+}
+
+/// <summary>
+/// A factory for creating <see cref="PixelMap{TPixel}"/> instances.
+/// </summary>
+internal static class PixelMapFactory
+{
+    /// <summary>
+    /// Creates a new <see cref="PixelMap{TPixel}"/> instance.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    /// <param name="configuration">The configuration.</param>
+    /// <param name="palette">The color palette to map from.</param>
+    /// <param name="colorMatchingMode">The color matching mode.</param>
+    /// <param name="transparentIndex">An explicit index at which to match transparent pixels.</param>
+    /// <returns>
+    /// The <see cref="PixelMap{TPixel}"/>.
+    /// </returns>
+    public static PixelMap<TPixel> Create<TPixel>(
+        Configuration configuration,
+        ReadOnlyMemory<TPixel> palette,
+        ColorMatchingMode colorMatchingMode,
+        int transparentIndex = -1)
+        where TPixel : unmanaged, IPixel<TPixel> => colorMatchingMode switch
+        {
+            ColorMatchingMode.Hybrid => new EuclideanPixelMap<TPixel, HybridCache>(configuration, palette, transparentIndex),
+            ColorMatchingMode.Exact => new EuclideanPixelMap<TPixel, NullCache>(configuration, palette, transparentIndex),
```

Comment:
This is a bit confusing. Does `ColorMatchingMode.Exact` mean that there is no caching? Shouldn't then we rename `ExactCache` to something like `AccurateCache`?

Reply:
Also, it might be good to add common `EuclideanPixelMap<TPixel, TCache>` instantiations to `AotCompilerTools`.

Reply:
Yeah, Exact means no cache. Very slow but the results are based on distance matching only.  I could rename the internal cache instance if that makes thing more clear.